[processing] Add flag to indicate whether an algorithm is safe

nyalldawson opened the pull request:
...and only run algorithms with this flag as background tasks.

Avoids issues with the Python GIL locking up the UI thread, by only allowing those algorithms which specifically flag that they can run in the background (i.e. the native c++ algorithms) to run as tasks.

It's possible other python based algs may NOT incur the GIL lock, and can also report this flag - but that will need to be done on an algorithm-by-algorithm basis.

TODO: for algorithms which can't be run in the background, we need to make sure the algorithm dialog is modal and prevents interaction with the QGIS window. I've needed to add processEvents calls to the dialog whenever the progress is changed in order to get the UI to refresh and detect cancellation requests, but we don't want this to allow interaction with the main qgis window.

nyalldawson committed:
[processing] Add flag to indicate whether an algorithm is safe
to run in a background thread

nyalldawson committed:
[processing] When running non-background enabled algorithms, use
processEvents to allow progress updates and cancelation clicks

nyalldawson committed:
Fix a possible crash in network analyser

nyalldawson committed:
Most c++ algorithms can run safely in background threads

alexbruy reviewed (approved):
+1 to merge.

Maybe this will encourage us to move more algorithms into core

nyalldawson committed:
Fix task manager UI sizes on hidpi displays

nyalldawson committed:
When an algorithm can run in the background, show this in the alg dialog

nyalldawson committed:
[processing] Show a modal progress dialog when running algorithms
which cannot run in background tasks

This is not fantastic UX, but we have lots of constraints here:

- The algorithm dialog itself cannot be made modal. There's child
widgets (such as the point and extent parameter widgets) which
interact with the main QGIS window.
- There is no reliable way in Qt to make a dialog modal after
it's shown (e.g. make it modal only when the algorithm is
running). Trust me - I've tried everything, and all approaches
break with some corner case.
- For non-background algorithms, we must have processEvents calls
in order to show the algorithm feedback and progress to users,
and detect cancel button clicks. Yet these processEvents calls
means that users can interact with other parts of QGIS, e.g.
removing layers from a project, and other operations which
could cause the algorithm to crash. So we MUST have some modal
dialog in order to block interactions outside of allowing
the cancel button clicks/progress repainting.

I've tried many approaches, but this is the only one which
works reliably...
